pip/compat/__init__.py

```diff
@@ -72,7 +72,11 @@ def console_to_str(s):
         try:
             return s.decode(sys.__stdout__.encoding)
         except UnicodeDecodeError:
-            return s.decode('utf_8')
+            try:
+                return s.decode('utf_8')
+            except UnicodeDecodeError:
+                # Decode at least line breaks manually
+                return repr(s)[2:-1].replace('\\r', '\r').replace('\\n', '\n')
```

Comment:
I'd be ready to merge that but I think it should emit some kind of warning to tell the user that pip is struggling with the terminal encoding.
Something like `warnings.warn("Could not detect stdout encoding - falling back to repr")`...